frontend/src/metabase/common/hooks/use-url-state.ts

```diff
@@ -0,0 +1,58 @@
+import type { Action, Location, Query } from "history";
+import { useCallback, useMemo } from "react";
+import { push, replace } from "react-router-redux";
+import { useEffectOnce } from "react-use";
+import { match } from "ts-pattern";
+
+import { useDispatch } from "metabase/lib/redux";
+
+export type QueryParam = Query[keyof Query];
+
+type BaseState = Record<string, unknown>;
+
+export type UrlStateConfig<State extends BaseState> = {
+  parse: (query: Query) => State;
+  serialize: (state: State) => Query;
+};
+
+type UrlStateActions<State extends BaseState> = {
+  patchUrlState: (
+    patch: Partial<State>,
+    action?: Extract<Action, "PUSH" | "REPLACE">,
+  ) => void;
+};
+
+/**
+ * Once we migrate to react-router 6 we should be able to replace this custom hook
+ * with something more sophisticated, like https://github.com/asmyshlyaev177/state-in-url
+ */
+export function useUrlState<State extends BaseState>(
```

Comment:
This hook gets slightly reworked in #56772 where a debounce is added to prevent adding browser history entries on every keystroke